crates/ruma-html/src/html.rs

```diff
@@ -402,6 +431,137 @@ pub struct ElementData {
+    pub fn data(&self) -> &'a NodeData {
+        &self.node.data
+    }
+
+    /// Returns the data of this node if it is a `NodeData::Element`.
+    pub fn as_element(&self) -> Option<&'a ElementData> {
+        self.node.as_element()
+    }
+
+    /// Returns the text content of this node, if it is a `NodeData::Text`.
+    pub fn as_text(&self) -> Option<&'a StrTendril> {
+        self.node.as_text()
+    }
+
+    /// The parent node of this node.
+    ///
+    /// Returns `None` if the parent is the root node.
+    pub fn parent(&self) -> Option<NodeRef<'a>> {
+        let parent_id = self.node.parent?;
+
+        // We don't want users to be able to navigate to the root.
+        if parent_id == self.html.root_id() {
+            return None;
+        }
+
+        Some(self.with_id(parent_id))
+    }
+
+    /// The next sibling node of this node.
+    ///
+    /// Returns `None` if this is the last of its siblings.
+    pub fn next_sibling(&self) -> Option<NodeRef<'a>> {
+        Some(self.with_id(self.node.next_sibling?))
+    }
+
+    /// The previous sibling node of this node.
+    ///
+    /// Returns `None` if this is the first of its siblings.
+    pub fn prev_sibling(&self) -> Option<NodeRef<'a>> {
+        Some(self.with_id(self.node.prev_sibling?))
+    }
+
+    /// Whether this node has children.
+    pub fn has_children(&self) -> bool {
+        self.node.first_child.is_some()
+    }
+
+    /// The first child node of this node.
+    ///
+    /// Returns `None` if this node has no children.
+    pub fn first_child(&self) -> Option<NodeRef<'a>> {
+        Some(self.with_id(self.node.first_child?))
+    }
+
+    /// The last child node of this node.
+    ///
+    /// Returns `None` if this node has no children.
+    pub fn last_child(&self) -> Option<NodeRef<'a>> {
+        Some(self.with_id(self.node.last_child?))
+    }
+
+    /// Get an iterator through the children of this node.
+    pub fn children(&self) -> Children<'a> {
+        Children::new(self.first_child())
+    }
+}
+
+/// An iterator through the children of a node.
+///
+/// Can be constructed with [`Html::children()`] or [`NodeRef::children()`].
+#[derive(Debug, Clone, Copy)]
+pub struct Children<'a> {
+    next: Option<NodeRef<'a>>,
+}
+
+impl<'a> Children<'a> {
+    /// Construct a `Children` starting from the given node.
+    fn new(start_node: Option<NodeRef<'a>>) -> Self {
+        Self { next: start_node }
+    }
+}
+
+impl<'a> Iterator for Children<'a> {
+    type Item = NodeRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.next?;
+        self.next = next.next_sibling();
+        Some(next)
+    }
+}
+
+impl<'a> FusedIterator for Children<'a> {}
+
+impl<'a> ExactSizeIterator for Children<'a> {
+    fn len(&self) -> usize {
+        let mut len = 0;
+
+        let mut next = self.next;
+        while let Some(node) = next {
+            len += 1;
+            next = node.next_sibling();
+        }
+
+        len
+    }
+}
```

Comment:
Is this impl actually helpful to you in practice? I find it rather surprising for an `ExactSizeIterator::len` implementation to have non-constant complexity.

Reply:
Probably not, I just know that its size is fixed so I wanted to implement the trait :stuck_out_tongue:.

Reply:
Haha, maybe let's leave it out for now then? Also, you implemented it wrong.

> When implementing an ExactSizeIterator, you must also implement [Iterator](https://doc.rust-lang.org/std/iter/trait.Iterator.html). When doing so, the implementation of [Iterator::size_hint](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint) _must_ return the exact size of the iterator.

> This method has a default implementation, so you usually should not implement it directly.

Reply:
Dropped